### PR TITLE
c++ encoder

### PR DESCRIPTION
**Pull request by catoverdrive**
builds on #4593, which I'm not sure how we're dealing with yet so I've left this unassigned for now.

@cseed

### CODE REVIEW

File: hail/src/main/scala/is/hail/cxx/TranslationUnit.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+class TranslationUnitBuilder() {
+
+  val definitions: ArrayBuilder[Definition] = new ArrayBuilder[Definition]()
+
+  val includes: ArrayBuilder[String] = new ArrayBuilder[String]()
+
+  def include(header: String): Unit = {
```
Comment (cseed) [inline hail/src/main/scala/is/hail/cxx/TranslationUnit.scala]: Option: `tub += Include("\"foo.h\"")`.  Or even `tub += SysInclude("vector")` and `tub += LocalInclude("foo.h")`.

File: hail/src/test/scala/is/hail/nativecode/NativeCodeSuite.scala [hunk 2]
```diff
@@ -218,4 +210,32 @@ class NativeCodeSuite extends SparkSuite {
     testPlus.close()
   }
 
+  @Test def testCXXCodeFunctions: Unit = {
+    val tub= new TranslationUnitBuilder()
```
Comment (cseed) [inline hail/src/test/scala/is/hail/nativecode/NativeCodeSuite.scala]: spacing.

File: hail/src/main/scala/is/hail/io/RowStore.scala [hunk 3]
```diff
@@ -1544,6 +1562,191 @@ final class PackEncoder(rowType: PType, out: OutputBuffer) extends Encoder {
...
+           |  $writeElt
+           |  $i++;
+           |  $eltOff += ${t.elementByteSize};
+           |}""".stripMargin)
+    )
+  }
+
+  def encodeBaseStruct(t: PBaseStruct, output_buf_ptr: cxx.Expression, off: cxx.Expression): cxx.Statement = {
+    //copy missing bytes
+    //iterate through non-missing fields
```
Comment (cseed) [inline hail/src/main/scala/is/hail/io/RowStore.scala]: space after `//` is traditional

File: hail/src/test/scala/is/hail/nativecode/NativeEncoderSuite.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import is.hail.cxx._
+import is.hail.expr.types._
+import is.hail.io._
+import org.testng.annotations.Test
+import is.hail.utils._
+import org.apache.spark.sql.Row
+
+class NativeEncoderSuite extends SparkSuite {
+
+  @Test def testCXXOutputStream: Unit = {
```
Comment (cseed) [inline hail/src/test/scala/is/hail/nativecode/NativeEncoderSuite.scala]: Unit functions should have `()`.

File: hail/src/main/resources/include/hail/Encoder.h [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+template <int BLOCKSIZE, typename OutputBlockBuffer>
+class LZ4OutputBlockBuffer {
+  private:
+    OutputBlockBuffer block_buf_;
+    char * block_;
+  public:
+    LZ4OutputBlockBuffer(std::shared_ptr<OutputStream> out) :
+      block_buf_(OutputBlockBuffer(out)),
+      block_(new char[LZ4_compressBound(BLOCKSIZE) + 4]{}) { }
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply hail/src/main/resources/include/hail/Encoder.h]: Do I need to call the destructor for `block_buf_` too, or will that be called automatically?

**Comment by cseed** [comment]
Code looks good.  It looks like there's something wrong with the gradle cpp test configuration.

**Pull request merged**